### PR TITLE
New resource/data source: `azurerm_mssql_database`

### PR DESCRIPTION
**Pull request by yupwei68**
fixes terraform-providers/terraform-provider-azurerm#2849
fixes terraform-providers/terraform-provider-azurerm#3978
fixes terraform-providers/terraform-provider-azurerm#1126
addresses #4185 

mssql db pass

### CODE REVIEW

File: azurerm/internal/services/mssql/data_source_mssql_database.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	mssqlServerId := d.Get("mssql_server_id").(string)
+	serverId, err := parse.MsSqlServerID(mssqlServerId)
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, serverId.ResourceGroup, serverId.Name, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			return fmt.Errorf("Error: Database %q (Resource Group %q, SQL Server %q) was not found", name, serverId.ResourceGroup, serverId.Name)
```
Comment (jackofallops) [inline azurerm/internal/services/mssql/data_source_mssql_database.go]:
```suggestion
			return fmt.Errorf("Database %q (Resource Group %q, SQL Server %q) was not found", name, serverId.ResourceGroup, serverId.Name)
```

File: azurerm/internal/services/mssql/data_source_mssql_database.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		}
+
+		return fmt.Errorf("Error making Read request on AzureRM Database %s (Resource Group %q, SQL Server %q): %+v", name, serverId.ResourceGroup, serverId.Name, err)
```
Comment (jackofallops) [inline azurerm/internal/services/mssql/data_source_mssql_database.go]:
```suggestion
		return fmt.Errorf("Failure making Read request on AzureRM Database %s (Resource Group %q, SQL Server %q): %+v", name, serverId.ResourceGroup, serverId.Name, err)
```

File: azurerm/internal/services/mssql/data_source_mssql_database.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+			Read: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"mssql_server_id": {
```
Comment (jackofallops) [inline azurerm/internal/services/mssql/data_source_mssql_database.go]: This could possibly be misleading as there is not an `azurerm_mssql_server` resource (yet)?

**Comment by yupwei68** [comment]
Hi @jackofallops Thanks for your comments. Corresponding changes have been pushed. Please continue to review the PR while you're available. Tests have passed.
mssql db pass2
datasource mssql db pass

File: azurerm/internal/services/mssql/validate/mssql_database_test.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+			input: "BC_S_Gen4_2",
+			valid: false,
+		},
+		{
+			name:  "Lower case",
+			input: "bc_gen5_2",
+			valid: true,
+		},
+	}
+	var validationFunction = ValidateMsSqlDBSkuName()
```
Comment (jackofallops) [inline azurerm/internal/services/mssql/validate/mssql_database_test.go]:
```suggestion
	var validationFunction = MsSqlDBSkuName()
```

File: azurerm/internal/services/mssql/validate/mssql_elastic_pool.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+	v, ok := i.(string)
+	if !ok {
+		errors = append(errors, fmt.Errorf("expected type of %q to be string", k))
+		return warnings, errors
+	}
+	if v == "" {
+		return warnings, errors
+	}
+
+	if _, _, _, err := parse.ParseArmMSSqlElasticPoolId(v); err != nil {
```
Comment (jackofallops) [inline azurerm/internal/services/mssql/validate/mssql_elastic_pool.go]:
```suggestion
	if _, _, _, err := parse.MSSqlElasticPoolId(v); err != nil {
```

File: website/docs/d/mssql_database.html.markdown [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+}
+```
+
+## Argument Reference
+
+* `name` - The name of the Ms SQL Database.
+
+* `sql_server_id` - The id of the Ms SQL Server on which to create the database.
+
+## Argument Reference
```
Comment (jackofallops) [inline website/docs/d/mssql_database.html.markdown]:
```suggestion
## Attribute Reference
```

File: website/docs/r/mssql_database.html.markdown [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+* `restore_point_in_time` - (Required) Specifies the point in time (ISO8601 format) of the source database that will be restored to create the new database. This property is only settable for `create_mode`= `PointInTimeRestore`  databases.
+
+* `read_replica_count` - (Optional) The number of readonly secondary replicas associated with the database to which readonly application intent connections may be routed. This property is only settable for Hyperscale edition databases.
+
+* `read_scale` - (Optional) If enabled, connections that have application intent set to readonly in their connection string may be routed to a readonly secondary replica. Possible values are `Enabled` and `Disabled`. This property is only settable for Premium and Business Critical databases.
+
+* `sample_name` - (Optional) Specifies the name of the sample schema to apply when creating this database. Possible value is `AdventureWorksLT`.
+
+* `sku_name` - (Optional) Specifies the name of the sku of the database. Changing this forces a new resource to be created. For example, `GP_Gen5_2`,`HS_Gen4_1`,`BC_Gen5_2`, `ElasticPool`, `Basic`,`S0`, `P2` , and etc.
```
Comment (jackofallops) [inline website/docs/r/mssql_database.html.markdown]: We should probably have a note here about the default value, since I believe it may be [different between regions](https://azure.microsoft.com/en-gb/updates/azure-sql-database-default-configuration-changing-soon/)? 
```suggestion
* `sku_name` - (Optional) Specifies the name of the sku of the database. Changing this forces a new resource to be created. For example, `GP_Gen5_2`,`HS_Gen4_1`,`BC_Gen5_2`, `ElasticPool`, `Basic`,`S0`, `P2`. 
~> **NOTE** The default sku_name value may differ between Azure locations depending on local availability of Gen4/Gen5 capacity.
```

File: website/docs/r/mssql_database.html.markdown [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+* `source_database_id` - (Required) The id of the source database to be copied to create the new database.
```
Comment (jackofallops) [inline website/docs/r/mssql_database.html.markdown]:
```suggestion
* `source_database_id` - (Optional) The id of the source database to be copied to create the new database. This should only be used for databases with `create_mode` values that use another database as reference.
```

File: website/docs/r/mssql_database.html.markdown [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the Ms SQL Database. Changing this forces a new resource to be created.
+
+* `sql_server_id` - (Required) The id of the Ms SQL Server on which to create the database. Changing this forces a new resource to be created.
```
Comment (jackofallops) [inline website/docs/r/mssql_database.html.markdown]: As mentioned above, possibly should be renamed to avoid confusion in future?  Also, possibly add a note to avoid confusion for "Serverless" still needing a Server resource to be associated to?
```suggestion
* `sql_server_id` - (Required) The id of the Ms SQL Server on which to create the database. Changing this forces a new resource to be created.
~> **NOTE:** This setting is still required for "Serverless" SKU's
```

File: azurerm/internal/services/mssql/resource_arm_mssql_database.go [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+		Read:   resourceArmMsSqlDatabaseRead,
+		Update: resourceArmMsSqlDatabaseCreateUpdate,
+		Delete: resourceArmMsSqlDatabaseDelete,
+		Importer: azSchema.ValidateResourceIDPriorToImport(func(id string) error {
+			_, err := parse.MsSqlDatabaseID(id)
+			return err
+		}),
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
```
Comment (jackofallops) [inline azurerm/internal/services/mssql/resource_arm_mssql_database.go]: This may need to be longer. The older database resource had 60 minutes, and I suspect that was due to some create_option values may take some time to complete?

**Comment by yupwei68** [comment]
Hi @jackofallops , thanks for comments with your experience. Corresponding changes are made. Not `forceNew` fields are tested in updates.  I have corrected my regular expression for `skuName` to include all the sku name available. I have added some checks like `source_database_id ` and `restore_point_in_time` in some `create_modes`. And as for your example, `min_capacity` and `auto_pause_delay` will be ignored in API in  other than `GP_S` databases, so I omits this check. Other checks like `read_replica_count` (settable under HS), `read_scale`, `zone_redundant`(settable under BC and Premium), they will be assigned default values in creation in all sku types, so during `terraform plan -refresh=false`, they might raise unexpected errors if the check of whether the fields are assigned  values under an unsuitable DB is added. So I prefer to leave this check to API during `terraform apply`. Tests are all passed. Please continue reviewing while you're available.

**Comment by yupwei68** [comment]
Hi @jackofallops Sorry for the errors made. And thanks for your comments. Corresponding changes are made. Please continue your review.

File: azurerm/internal/services/mssql/resource_arm_mssql_database.go [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+					string(sql.BasePrice),
+					string(sql.LicenseIncluded),
+				}, false),
+			},
+
+			"max_size_gb": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validation.IntBetween(1, 1024),
```
Comment (jackofallops) [inline azurerm/internal/services/mssql/resource_arm_mssql_database.go]: I think these values may be too restrictive?  Different vCore / DTU values have different limits, so I think a basic check is still the best option here, but update the docs to point the user at documentation resources for the value, such as: https://docs.microsoft.com/en-us/azure/sql-database/sql-database-service-tiers-general-purpose-business-critical
```suggestion
				ValidateFunc: validation.IntBetween(1, 4096),
```

Note: we may need to change it to `IntAtLeast(1)` in future if the service increases the upper boundary.

File: azurerm/internal/services/mssql/resource_arm_mssql_database.go [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_mssql_database", *existing.ID)
+		}
+	}
+
+	serverClient := meta.(*clients.Client).MSSQL.ServersClient
+	serverResp, err := serverClient.Get(ctx, serverId.ResourceGroup, serverId.Name)
+	if err != nil {
+		return fmt.Errorf("Failure in making Read request on MsSql Server  %q (Resource Group %q): %s", serverId.Name, serverId.ResourceGroup, err)
```
Comment (jackofallops) [inline azurerm/internal/services/mssql/resource_arm_mssql_database.go]: whitespace
```suggestion
		return fmt.Errorf("Failure in making Read request on MsSql Server %q (Resource Group %q): %s", serverId.Name, serverId.ResourceGroup, err)
```

File: azurerm/internal/services/mssql/validate/mssql_database.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+func MsSqlDatabaseAutoPauseDelay(i interface{}, k string) (warnings []string, errors []error) {
+	v, ok := i.(int)
+	if !ok {
+		errors = append(errors, fmt.Errorf("expected type of %s to be integer", k))
+		return warnings, errors
+	}
+	min := 60
+	max := 10080
+	if (v < min || v > max) && v%10 != 0 && v != -1 {
+		errors = append(errors, fmt.Errorf("expected %s to be in the range (%d - %d) and divisible by 10 or -1, got %d", k, min, max, v))
```
Comment (jackofallops) [inline azurerm/internal/services/mssql/validate/mssql_database.go]: This shouldn't be able to be `-1`?
```suggestion
		errors = append(errors, fmt.Errorf("expected %s to be in the range (%d - %d) and divisible by 10, got %d", k, min, max, v))
```

File: website/docs/r/mssql_database.html.markdown [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the Ms SQL Database. Changing this forces a new resource to be created.
+
+* `server_id` - (Required) The id of the Ms SQL Server on which to create the database. Changing this forces a new resource to be created.
+~> **NOTE:** This setting is still required for "Serverless" SKU's
```
Comment (jackofallops) [inline website/docs/r/mssql_database.html.markdown]: markdown formatting
```suggestion
* `server_id` - (Required) The id of the Ms SQL Server on which to create the database. Changing this forces a new resource to be created.

~> **NOTE:** This setting is still required for "Serverless" SKU's
```

File: website/docs/r/mssql_database.html.markdown [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+* `read_replica_count` - (Optional) The number of readonly secondary replicas associated with the database to which readonly application intent connections may be routed. This property is only settable for Hyperscale edition databases.
+
+* `read_scale` - (Optional) If enabled, connections that have application intent set to readonly in their connection string may be routed to a readonly secondary replica. Possible values are `Enabled` and `Disabled`. This property is only settable for Premium and Business Critical databases.
+
+* `sample_name` - (Optional) Specifies the name of the sample schema to apply when creating this database. Possible value is `AdventureWorksLT`.
+
+* `sku_name` - (Optional) Specifies the name of the sku of the database. Changing this forces a new resource to be created. For example, `GP_S_Gen5_2`,`HS_Gen4_1`,`BC_Gen5_2`, `ElasticPool`, `Basic`,`S0`, `P2` ,`DW100c`, `DS100`.
+~> **NOTE** The default sku_name value may differ between Azure locations depending on local availability of Gen4/Gen5 capacity.
```
Comment (jackofallops) [inline website/docs/r/mssql_database.html.markdown]: markdown formatting
```suggestion
* `sku_name` - (Optional) Specifies the name of the sku of the database. Changing this forces a new resource to be created. For example, `GP_S_Gen5_2`,`HS_Gen4_1`,`BC_Gen5_2`, `ElasticPool`, `Basic`,`S0`, `P2` ,`DW100c`, `DS100`.

~> **NOTE** The default sku_name value may differ between Azure locations depending on local availability of Gen4/Gen5 capacity.
```

File: azurerm/internal/services/mssql/validate/mssql_database.go (hunk 13 above)
Comment (yupwei68) [reply azurerm/internal/services/mssql/validate/mssql_database.go]: In the API, they suppose `-1` as disable `auto pause delay`. So I explain it in the resource doc:`A value of -1 means that automatic pause is disabled`.

**Comment by yupwei68** [comment]
Hi @jackofallops Thanks for the comments. Corresponding changes are updated, except one unresolved comment. Looking forward to get merged, too.

File: azurerm/internal/services/mssql/tests/resource_arm_mssql_database_test.go [hunk 16]
```diff
@@ -0,0 +1,634 @@
...
+}
+
+func testAccAzureRMMsSqlDatabase_template(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-mssqldb-%[1]d"
```
Comment (katbyte) [inline azurerm/internal/services/mssql/tests/resource_arm_mssql_database_test.go]: can we shorten this to just the service abbreviation 
```suggestion
  name     = "acctestRG-mssql-%[1]d"
```

File: website/docs/r/mssql_database.html.markdown [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the Ms SQL Database. Changing this forces a new resource to be created.
+
+* `server_id` - (Required) The id of the Ms SQL Server on which to create the database. Changing this forces a new resource to be created.
+
+~> **NOTE:** This setting is still required for "Serverless" SKU's
+
+* `auto_pause_delay_in_minutes` - (Optional) Time in minutes after which database is automatically paused. A value of -1 means that automatic pause is disabled. This property is only settable for General Purpose Serverless databases.
```
Comment (katbyte) [inline website/docs/r/mssql_database.html.markdown]: could we also include it needs to be divisible by 10?
```suggestion
* `auto_pause_delay_in_minutes` - (Optional) Time in minutes after which database is automatically paused. A value of `-1` means that automatic pause is disabled. This property is only settable for General Purpose Serverless databases.
```

File: azurerm/internal/services/mssql/validate/mssql_database.go [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+		return warnings, errors
+	}
+
+	return warnings, errors
+}
+
+func MsSqlDBMinCapacity(i interface{}, k string) (warnings []string, errors []error) {
```
Comment (katbyte) [inline azurerm/internal/services/mssql/validate/mssql_database.go]: could we just write a `validate.FloatInSlice` and put it into helpers for this instead of a custom validation function?

**Comment by yupwei68** [comment]
Hi @katbyte , Thanks for your review comments. Corresponding changes are updated. All tests have been retested with pass.

File: azurerm/internal/services/mssql/resource_arm_mssql_database.go [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+
+			"read_scale": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(sql.DatabaseReadScaleEnabled),
+					string(sql.DatabaseReadScaleDisabled),
+				}, false),
+			},
```
Comment (katbyte) [inline azurerm/internal/services/mssql/resource_arm_mssql_database.go]: Given this is a boolean enabled/disabled property could we make it TypeBool?
```suggestion
			"read_scale_enabled": {
				Type:     schema.TypeBool,
				Optional: true,
				Computed: true,
			},
```

File: azurerm/internal/services/mssql/resource_arm_mssql_database.go [hunk 20]
```diff
@@ -0,0 +1,363 @@
...
+			"sku_name": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				ForceNew:         true,
+				Computed:         true,
+				ValidateFunc:     validate.MsSqlDBSkuName(),
+				DiffSuppressFunc: suppress.CaseDifference,
+			},
+
+			"source_database_id": {
```
Comment (katbyte) [inline azurerm/internal/services/mssql/resource_arm_mssql_database.go]: Would this make more sense as
```suggestion
			"creation_source_database_id": {
```

File: website/docs/d/mssql_database.html.markdown [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+* `name` - The name of the Ms SQL Database.
+
+* `server_id` - The id of the Ms SQL Server on which to create the database.
+
+## Attribute Reference
+
+* `collation` - The collation of the database. 
+
+* `elastic_pool_id` - The id of the elastic pool containing this database.
```
Comment (katbyte) [inline website/docs/d/mssql_database.html.markdown]:
```suggestion
* `elastic_pool_id` - The ID of the elastic pool containing this database.
```

File: website/docs/d/mssql_database.html.markdown [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+
+* `license_type` - The license type to apply for this database.
```
Comment (katbyte) [inline website/docs/d/mssql_database.html.markdown]:
```suggestion
* `license_type` - The license type applied to this database.
```

File: website/docs/d/mssql_database.html.markdown [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+
+* `max_size_gb` - The max size of the database in gigabytes.
+
+* `read_replica_count` - The number of readonly secondary replicas associated with the database to which readonly application intent connections may be routed. 
+
+* `read_scale` - If enabled, connections that have application intent set to readonly in their connection string may be routed to a readonly secondary replica.
+
+* `sku_name` - The name of the sku of the database.
```
Comment (katbyte) [inline website/docs/d/mssql_database.html.markdown]:
```suggestion
* `sku_name` - The name of the sku used by the database.
```

**Comment by yupwei68** [comment]
Hi @katbyte ,Sure! Thanks. Please continue your review.
